doc/neps/nep-0018-array-function-protocol.rst

```diff
@@ -250,87 +306,116 @@ an ``__array_function__`` method that is identical to
+``array_function_dispatch`` decorator, which also associates dispatcher
+functions:
 
 .. code:: python
 
-    @overload_for_array_function
-    def broadcast_to(array: ArrayLike
-                     shape: Tuple[int, ...],
-                     subok: bool = False):
-        ...  # continue with the definition of broadcast_to
-
-The decorator ``overload_for_array_function`` would be written in terms
-of ``try_array_function_override``, but would also need some level of magic
-for (1) access to the wrapper function (``np.broadcast_to``) for passing into
-``__array_function__`` implementations and (2) dynamic code generation
-resembling the `decorator library <https://github.com/micheles/decorator>`_ 
-to automatically write an overloaded function like the manually written
-implemenations above with the exact same signature as the original.
-Unfortunately, using the ``inspect`` module instead of code generation would
-probably be too slow: our prototype implementation adds ~15 microseconds of
-overhead.
-
-We like the idea of writing overloads with minimal syntax, but dynamic
-code generation also has potential downsides, such as slower import times, less
-transparent code and added difficulty for static analysis tools. It's not clear
-that tradeoffs would be worth it, especially because functions with complex
-signatures like ``np.einsum`` would assuredly still need to invoke
-``try_array_function_override`` directly.
-
-So we don't propose adding such a decorator yet, but it's something worth
-considering for the future.
+    def array_function_dispatch(dispatcher):
+        """Wrap a function for dispatch with the __array_function__ protocol."""
+        def decorator(func):
+            @functools.wraps(func)
+            def new_func(*args, **kwargs):
+                relevant_arguments = dispatcher(*args, **kwargs)
+                success, value = try_array_function_override(
+                    new_func, relevant_arguments, args, kwargs)
+                if success:
+                    return value
+                return func(*args, **kwargs)
+            return new_func
+        return decorator
+
+    # example usage
+    def _broadcast_to_dispatcher(array, shape, subok=None, **ignored_kwargs):
+        return (array,)
+
+    @array_function_dispatch(_broadcast_to_dispatcher)
+    def broadcast_to(array, shape, subok=False):
+        ...  # existing definition of np.broadcast_to
+
+Using a decorator is great! We don't need to change the definitions of
+existing NumPy functions, and only need to write a few additional lines
+for the dispatcher function. We could even reuse a single dispatcher for
+families of functions with the same signature (e.g., ``sum`` and ``prod``).
+For such functions, the largest change could be adding a few lines to the
+docstring to note which arguments are checked for overloads.
+
+It's particularly worth calling out the decorator's use of
+``functools.wraps``:
```

Comment:
It might be worth mentioning that this isn't  a catch-all thing, it *does* sometimes not work on arbitrary callables if they aren't Python functions. I ran into this trying to nicely wrap ufuncs. [Here's a quick demo](https://gist.github.com/hameerabbasi/7301e2895f5540745daa65365b5a8890), it shows that `repr` and `string` don't work, but `__doc__` does. It's worth noting that this can be worked around, see https://github.com/pydata/sparse/blob/b51d74924d62ff6537b15ce4e1dd4e56080a3b6f/sparse/utils.py#L187-L191

It might be worth investigating this as a Python issue, but I'm not well-versed enough in Python to say if this is actually a bug.

Reply:
Yeah, I think this only works well Python functions. For classes/methods, we would need another solution.

But I think functions are mostly what we want to cover in this NEP.